### PR TITLE
consensus/istanbul: implement proof of consensus

### PR DESCRIPTION
**Pull request by bailantaotao**

### CODE REVIEW

File: core/types/istanbul.go [hunk 1]
```diff
@@ -30,15 +31,127 @@ var (
...
 	IstanbulExtraSeal          = 65                               // Fixed number of extra-data suffix bytes reserved for signer seal
+	IstanbulExtraCommittedSize = 1                                // Fixed number of extra-data infix bytes reserved for committed size
+	IstanbulExtraCommittedSeal = 65                               // Fixed number of extra-data suffix bytes reserved for committed seal
+
+	ErrInvalidIstanbulHeaderExtra   = fmt.Errorf("Invalid istanbul header extra-data")
+	ErrInvalidIstanbulCommittedSeal = fmt.Errorf("Invalid istanbul committed seal")
 )
 
+// FIXME: I don't like this name :(
+type Istanbul struct {
```
Comment (bailantaotao) [inline core/types/istanbul.go]: IstanbulExtra

File: consensus/istanbul/backends/simple/backend.go [hunk 2]
```diff
@@ -125,10 +130,15 @@ func (sb *simpleBackend) Commit(proposal istanbul.Proposal) error {
 	// -- otherwise, a error will be returned and a round change event will be fired.
 	if sb.proposedBlockHash == block.Hash() {
 		// feed block hash to Seal() and wait the Seal() result
-		sb.commitCh <- block.Hash()
+		sb.commitCh <- committedResult{Hash: block.Hash(), Signatures: signatures}
 		// TODO: how do we check the block is inserted correctly?
 		return nil
 	}
+
+	err := block.UpdateIstanbulHeaderExtra(signatures)
```
Comment (bailantaotao) [inline consensus/istanbul/backends/simple/backend.go]: use ustalbul.UpdateHeaderExtra

File: consensus/istanbul/backends/simple/backend.go [hunk 3]
```diff
@@ -42,11 +42,16 @@ func New(config *istanbul.Config, eventMux *event.TypeMux, privateKey *ecdsa.Pri
 		address:          crypto.PubkeyToAddress(privateKey.PublicKey),
 		logger:           log.New("backend", "simple"),
 		db:               db,
-		commitCh:         make(chan common.Hash, 1),
+		commitCh:         make(chan committedResult, 1),
```
Comment (bailantaotao) [inline consensus/istanbul/backends/simple/backend.go]: types.Block

File: core/types/istanbul.go [hunk 4]
```diff
@@ -30,15 +31,127 @@ var (
...
+
+func ExtractToIstanbul(h *Header) *Istanbul {
+	index := ExtractToIstanbulIndex(h)
+
+	validators := make([]common.Address, (index.Seal-index.ValidatorLength)/common.AddressLength)
+	for i := 0; i < len(validators); i++ {
+		copy(validators[i][:], h.Extra[index.ValidatorLength+i*common.AddressLength:])
+	}
+
+	cmttedSeals := make([][]byte, (len(h.Extra)-index.CommittedSealLength)/IstanbulExtraSeal)
```
Comment (yutelin) [inline core/types/istanbul.go]: IstanbulExtraCommittedSeal

File: core/types/istanbul.go [hunk 5]
```diff
@@ -30,15 +31,127 @@ var (
...
+	for i := 0; i < len(cmttedSeals); i++ {
+		cmttedSeals[i] = make([]byte, IstanbulExtraCommittedSeal)
+		copy(cmttedSeals[i][:], h.Extra[index.CommittedSealLength+i*IstanbulExtraSeal:])
```
Comment (yutelin) [inline core/types/istanbul.go]: IstanbulExtraCommittedSeal

File: core/types/istanbul.go [hunk 6]
```diff
@@ -30,15 +31,127 @@ var (
...
+	// a sanity check has done before.
+	valLength := getValidatorLength(header)
+	if len(header.Extra) < defaultLength+valLength {
+		return ErrInvalidIstanbulHeaderExtra
+	}
+
+	return nil
+}
+
+// PrepareIstanbulExtra returns a istanbul extra-data that use the given header, parent
```
Comment (yutelin) [inline core/types/istanbul.go]: Might need to rephrase the comment.

File: core/types/istanbul.go [hunk 7]
```diff
@@ -30,15 +31,127 @@ var (
...
+// we copy vanity from header and validator, seal from parent
+func PrepareIstanbulExtra(header, parent *Header) []byte {
+	var buf bytes.Buffer
+
```
Comment (yutelin) [inline core/types/istanbul.go]: Comment the following steps.

File: core/types/istanbul.go [hunk 8]
```diff
@@ -30,15 +31,127 @@ var (
...
+	index := ExtractToIstanbulIndex(parent)
+
+	buf.Write(parent.Extra[index.ValidatorSize:index.Seal])
+	buf.Write(make([]byte, IstanbulExtraSeal))
+
+	return buf.Bytes()
+}
+
 // IstanbulExtraFilter is used to filter out the useless information.
```
Comment (yutelin) [inline core/types/istanbul.go]: useless information for what?

File: core/types/istanbul.go [hunk 9]
```diff
@@ -30,15 +31,127 @@ var (
 	IstanbulExtraVanity        = int(params.MaximumExtraDataSize) // Fixed number of extra-data prefix bytes reserved for signer vanity
 	IstanbulExtraValidatorSize = 1                                // Fixed number of extra-data infix bytes reserved for validator size
 	IstanbulExtraSeal          = 65                               // Fixed number of extra-data suffix bytes reserved for signer seal
+	IstanbulExtraCommittedSize = 1                                // Fixed number of extra-data infix bytes reserved for committed size
+	IstanbulExtraCommittedSeal = 65                               // Fixed number of extra-data suffix bytes reserved for committed seal
```
Comment (yutelin) [inline core/types/istanbul.go]: probably just remove `IstanbulExtraCommittedSeal` and use `IstanbulExtraSeal` directly?

Comment (yutelin) [inline core/types/istanbul.go]: not sure about this? any suggestion?

File: core/types/istanbul.go [hunk 10]
```diff
@@ -49,16 +162,46 @@ func IstanbulExtraFilter(h *Header) *Header {
...
+		newHeader.Extra = append(newHeader.Extra, bytes.Repeat([]byte{0x00}, defaultLength-len(newHeader.Extra))...)
+	}
+
+	newHeader.Extra = newHeader.Extra[0:defaultLength]
 
 	return newHeader
 }
+
+// updateIstanbulCommittedSeal updates header signatures that store consensus proof
+func updateIstanbulCommittedSeal(h *Header, committedSeal []byte) error {
```
Comment (yutelin) [inline core/types/istanbul.go]: `appendIstanbulCommittedSeal` ?

Comment (yutelin) [inline core/types/istanbul.go]: not a strong opinion. you make the call.

File: core/types/istanbul.go [hunk 11]
```diff
@@ -49,16 +162,46 @@ func IstanbulExtraFilter(h *Header) *Header {
...
+	size := len(committedSeal) / IstanbulExtraCommittedSeal
+	h.Extra = append(h.Extra, byte(size))
+	h.Extra = append(h.Extra, committedSeal...)
+	return nil
+}
+
+// getValidatorLength returns a validator total length for a given header
+func getValidatorLength(h *Header) int {
+	if len(h.Extra) < IstanbulExtraVanity+IstanbulExtraValidatorSize {
+		return -1
```
Comment (yutelin) [inline core/types/istanbul.go]: 0

File: core/types/istanbul.go (hunk 4 above)
Comment (bailantaotao) [inline core/types/istanbul.go]: done

File: core/types/istanbul.go (hunk 6 above)
Comment (bailantaotao) [inline core/types/istanbul.go]: done

File: core/types/istanbul.go (hunk 7 above)
Comment (bailantaotao) [inline core/types/istanbul.go]: done

File: core/types/istanbul.go (hunk 8 above)
Comment (bailantaotao) [inline core/types/istanbul.go]: done

File: core/types/istanbul.go (hunk 9 above)
Comment (bailantaotao) [inline core/types/istanbul.go]: 我最後只留IstanbulExtraSeal，因為簽名出來的長度就是65，不管怎樣都不會變動，且committed
 seals也是seal組成，所以....

File: core/types/istanbul.go (hunk 10 above)
Comment (bailantaotao) [inline core/types/istanbul.go]: done

File: core/types/istanbul.go (hunk 11 above)
Comment (bailantaotao) [inline core/types/istanbul.go]: done

File: core/types/istanbul.go [hunk 12]
```diff
@@ -27,18 +28,134 @@ var (
...
+	IstanbulExtraCommittedSize = 1                                // Fixed number of extra-data bytes reserved for committed size
+
+	ErrInvalidIstanbulHeaderExtra   = fmt.Errorf("Invalid istanbul header extra-data")
+	ErrInvalidIstanbulCommittedSeal = fmt.Errorf("Invalid istanbul committed seal")
 )
 
-// IstanbulExtraFilter is used to filter out the useless information.
-//
-// We just keep the following data in header.Extra.
+type IstanbulExtra struct {
```
Comment (markya0616) [inline core/types/istanbul.go]: Can we implement rlp encode/decode? In my imagination, it should be more clean and simple.

File: core/types/istanbul.go [hunk 13]
```diff
@@ -27,18 +28,134 @@ var (
 	// The IstanbulDigest represents a constant of "Istanbul practical byzantine fault tolerance".
 	IstanbulDigest = common.HexToHash("0x63746963616c2062797a616e74696e65206661756c7420746f6c6572616e6365")
 
-	IstanbulExtraVanity        = int(params.MaximumExtraDataSize) // Fixed number of extra-data prefix bytes reserved for signer vanity
-	IstanbulExtraValidatorSize = 1                                // Fixed number of extra-data infix bytes reserved for validator size
-	IstanbulExtraSeal          = 65                               // Fixed number of extra-data suffix bytes reserved for signer seal
+	IstanbulExtraVanity        = int(params.MaximumExtraDataSize) // Fixed number of extra-data bytes reserved for signer vanity
```
Comment (markya0616) [inline core/types/istanbul.go]: use constant value

File: core/types/istanbul.go (hunk 12 above)
Comment (bailantaotao) [inline core/types/istanbul.go]: 1. 如果extra-data是儲存rlp編碼結果的話，那我們的genesis需要額外做一個tool，去產生包含validator，且經過rlp編碼的資料給extra-data。
2. block做簽名時需要包含validator，這樣在做簽名時，我們必須先解碼extra-data，然後將某些欄位設成空。
3. 當初我沒有實做rlp encode/decode主要是因為validator也要一起被簽名，然後不想額外再做一個tool去產生genesis extra-data

但我承認，透過rlp實作的話會比較簡單。

**Pull request merged**